typescript/infra/scripts/helloworld/kathy.ts

```diff
@@ -54,51 +54,79 @@ async function main() {
-  const metricsInterval = setInterval(() => {
-    submitMetrics(metricsRegister, 'kathy', { appendMode: true }).catch(
-      console.error,
-    );
-  }, 1000 * 30);
-
-  for (const origin of origins) {
-    for (const destination of origins.filter((d) => d !== origin)) {
-      const labels = {
+  // track how many we are still allowed to send in case some messages send slower than expected.
+  let allowedToSend = 0;
+  setInterval(() => {
+    allowedToSend++;
+  }, fullCycleTime / pairings.length);
+
+  for (
+    // in case we are restarting kathy, keep it from always running the exact same messages first
+    let currentPairingIndex = Date.now() % pairings.length;
+    ;
+    currentPairingIndex = (currentPairingIndex + 1) % pairings.length
+  ) {
+    currentPairingIndexGauge.set(currentPairingIndex);
+    // wait until we are allowed to send the message
+    while (allowedToSend <= 0) await sleep(1000);
+    allowedToSend--;
+
+    const { origin, destination } = pairings[currentPairingIndex];
+    const labels = {
+      origin,
+      remote: destination,
+    };
+    try {
+      await sendMessage(app, origin, destination, gasCalc);
+      log('Message sent successfully', { origin, destination });
+      messagesSendCount.labels({ ...labels, status: 'success' }).inc();
+    } catch (e) {
+      error(`Error sending message, continuing...`, {
+        error: e,
         origin,
-        remote: destination,
-        ...constMetricLabels,
-      };
-      try {
-        await sendMessage(app, origin, destination, gasCalc);
-        messagesSendStatus.labels({ ...labels }).set(1);
-      } catch (err) {
-        console.error(
-          `Error sending message from ${origin} to ${destination}, continuing...`,
-          `${err}`.replaceAll('\n', ' ## '),
-        );
-        failureOccurred = true;
-        messagesSendStatus.labels({ ...labels }).set(0);
-      }
-
-      // Sleep 500ms to avoid race conditions where nonces are reused
-      await sleep(500);
+        destination,
+      });
+      messagesSendCount.labels({ ...labels, status: 'failure' }).inc();
     }
-  }
-
-  clearInterval(metricsInterval);
-  // do not use append mode here so we can clear any old pairings we no longer care about.
-  await submitMetrics(metricsRegister, 'kathy', { appendMode: false });
 
-  if (failureOccurred) {
-    console.error('Failure occurred at least once');
-    process.exit(1);
+    // print stats once every cycle through the pairings
+    if (currentPairingIndex == 0) {
+      for (const [origin, destinationStats] of Object.entries(
+        await app.stats(),
+      )) {
+        for (const [destination, counts] of Object.entries(destinationStats)) {
+          debug('Message stats', { origin, destination, ...counts });
+        }
+      }
+    }
   }
 }
 
+// 10 min by default
+const MESSAGE_RECEIPT_TIMEOUT =
+  parseInt(process.env['KATHY_MESSAGE_RECEIPT_TIMEOUT'] as string) ||
+  10 * 60 * 1000;
```

Comment:
I'm a little worried this is cutting it a little close for polygon, where 256 blocks to finality * ~2 second block times = 8.5 min, which only gives 1.5 mins for delivery. I know at the hackathon I saw messages take > 10 min to be delivered from Polygon -> Ethereum, sometimes because it also can take a while for a tx to be included in a block on Mumbai. As we're still early and aren't worrying too much about message latency at this point, maybe something like 15 min is best? And then when we have the time to more carefully care about latency we can wean this down. Also I realize this is just the default and the 15 min should/could just be configured at the infra level, just got thinking of this here

This also makes me wonder if we should actually track the time for a message sent by kathy to be delivered, which could provide useful insights. Right now I don't think we have any direct tracking of this, wdyt? Happy for it to just be something in the backlog for the future, not necessarily saying we should add that in this PR

Reply:
Yep, good metric idea, will add.